### PR TITLE
Sampling signature upgrades

### PR DESCRIPTION
**Pull request by positr0nium**
## Description

Refactors the return signature of `sample()` and `expectation_value()` so that
the output structure mirrors the kernel's return structure.  Instead of
squeezing multi-value returns into a flat 2D array (which forced a common
`float64` dtype), each leaf of the return pytree now gets its own typed 1D
accumulator.  Nested tuples, lists, and dicts are preserved in the result.

## Related Issues

Related to the `backend_sampling` feature set.

## Type of Change

- [x] Feature (new functionality)
- [ ] Change Request (modification of existing functionality)
- [ ] Bug Fix
- [ ] Refactoring (no behavior change)
- [ ] Performance improvement
- [ ] Documentation
- [ ] CI / Build

## Breaking Change?
- [x] Yes
- [ ] No

If yes, describe the impact and migration path:

Multi-value returns from `sample()` now produce a **tuple of 1D arrays**
instead of a single 2D `(shots, n)` array.  Code that indexed `result[:, i]`
must be updated to use `result[i]`.  The same applies to `expectation_value()`
which now returns a **tuple of scalars** instead of a 1D array.  See the
migration notes below.

## What was changed?

- **`sampling.py`**: Replaced the single flat 2D accumulator with a tuple of
  typed per-leaf 1D accumulators.  Added `_update_acc()` / `_make_init_acc()`
  helpers that capture the pytree structure via `jax.tree_util` on the first
  iteration, restart the loop with correctly-shaped accumulators, and
  reconstruct the nested structure via `tree_unflatten` after the loop.
  Array-valued leaves (e.g. `jnp.array([1,2,3])`) are supported — their inner
  dimensions become leading accumulator dimensions `(shots, 3)`.
  `shots=0` raises `ValueError` early in the JIT path.
  User-defined pytree types raise `TypeError` with a clear message.

- **`ev.py`**: Same pytree accumulator strategy applied to
  `expectation_value`.  Running-sum accumulators (not per-shot storage)
  preserve leaf shapes.  The legacy `dict_sampling_eval_function` path (used
  by terminal sampling) keeps the old flat-array format for compatibility.
  The identity post-processor now unwraps single-value returns (consistent
  with `sample()`).

- **`backend_sampling.py`**: `_make_backend_eqn_evaluator` handles multiple
  `eqn.outvars` so `pure_callback` can consume the new pytree output.

- **`backend_sampling_interpreter.py`**: `backend_sampling_fn` flattens
  nested pytree results for `pure_callback` compatibility.

- **`terminal_sampling_interpreter.py`**: Splits flat "array" results (2D →
  1D arrays) and "ev" results (1D → 0D scalars) when the Jaxpr has multiple
  outvars.

- **Tests**: Added 71 new tests across `test_sampling_return_signature.py`
  (44 tests) and `test_ev_return_signature.py` (27 tests), covering scalars,
  flat/nested tuples, lists, dicts, dtype preservation, array-valued leaves,
  post-processor variants, edge cases, and stress scenarios.  Updated 6
  existing tests for the new tuple format.

## How was it tested?

All 125 relevant tests pass (54 original + 71 new).

| Test suite | Tests | Result |
|---|---|---|
| `test_sampling.py` | 4 | ✅ |
| `test_backend_sampling.py` | 32 | ✅ |
| `test_backend_sampling_ev.py` | 18 | ✅ |
| `test_sampling_return_signature.py` | 44 | ✅ |
| `test_ev_return_signature.py` | 27 | ✅ |

### Migration examples

**sample() — before:**
```python
res = sample(kernel, shots=100)(args)
# res.shape == (100, 3)  — single 2D array
first_col = res[:, 0]
```

**sample() — after:**
```python
res = sample(kernel, shots=100)(args)
# res == (array1, array2, array3)  — tuple of three 1D arrays
first_col = res[0]
```

**expectation_value() — before:**
```python
res = expectation_value(kernel, shots=500)(args)
# res.shape == (3,)  — 1D array of means
```

**expectation_value() — after:**
```python
res = expectation_value(kernel, shots=500)(args)
# res == (mean1, mean2, mean3)  — tuple of scalars
```

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [x] I have added/updated tests
- [ ] All tests pass locally
- [x] I have updated the documentation (docstrings updated)
- [x] I have added a changelog entry
- [x] Breaking changes are documented with migration path

### CODE REVIEW

**Comment by PietropaoloFrisoni** [comment]
Thanks @positr0nium , I should be able to look into this tomorrow.

In the meantime, I will request Copilot's review (I hope you don't mind) since it is quite a useful tool for my PRs

File: documentation/source/general/changelog/changelog-dev.rst [hunk 1]
```diff
@@ -10,6 +10,23 @@ changelog enforcement.
...
+  ``expectation_value()``
+  The return structure now mirrors the structure of the sampling kernel's
+  return value.  For example ``return a, b, c`` from a kernel produces
+  ``(array_a, array_b, array_c)`` — a tuple of three 1D arrays — instead of
+  a single flat ``(shots, 3)`` array.  Nested tuples, lists, and dicts are
+  preserved (e.g. ``[ (a,b), c ]`` → ``[ (array_a, array_b), array_c ]``).
+  Each leaf retains its native dtype (``bool`` stays ``bool``, ``int`` stays
+  ``int``) and array-valued leaves naturally stack along the leading
+  dimension (``(3,)`` → ``(shots, 3)``).  User-defined JAX pytree types
+  raise a descriptive ``TypeError``.
```
Comment (PietropaoloFrisoni) [inline documentation/source/general/changelog/changelog-dev.rst]:
```suggestion
  raise a descriptive ``TypeError``.
  (`PR #749 <https://github.com/eclipse-qrisp/Qrisp/pull/749>`_)
```

File: documentation/source/general/changelog/changelog-dev.rst [hunk 2]
```diff
@@ -10,6 +10,23 @@ changelog enforcement.
...
+
+  **Breaking change:** Multi-value returns are now tuples of arrays instead
+  of a single 2D array.  Code that indexed ``res[:, i]`` must use
+  ``res[i]``.  The same applies to ``expectation_value()`` which now returns
+  a tuple of scalars for multi-value kernels.
```
Comment (PietropaoloFrisoni) [inline documentation/source/general/changelog/changelog-dev.rst:32]:
```suggestion
  a tuple of scalars for multi-value kernels.
  (`PR #749 <https://github.com/eclipse-qrisp/Qrisp/pull/749>`_)
```

File: tests/jax_tests/test_ev_return_signature.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+*
+* This Source Code may also be made available under the following Secondary
+* Licenses when the conditions for such availability set forth in the Eclipse
+* Public License, v. 2.0 are satisfied: GNU General Public License, version 2
+* with the GNU Classpath Exception which is
+* available at https://www.gnu.org/software/classpath/license.html.
+*
+* SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+********************************************************************************
+"""
```
Comment (PietropaoloFrisoni) [inline tests/jax_tests/test_ev_return_signature.py]:
```suggestion
# ********************************************************************************
# * Copyright (c) 2026 the Qrisp authors
# *
# * This program and the accompanying materials are made available under the
# * terms of the Eclipse Public License 2.0 which is available at
# * http://www.eclipse.org/legal/epl-2.0.
# *
# * This Source Code may also be made available under the following Secondary
# * Licenses when the conditions for such availability set forth in the Eclipse
# * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
# * with the GNU Classpath Exception which is
# * available at https://www.gnu.org/software/classpath/license.html.
# *
# * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
# ********************************************************************************
```

File: tests/jax_tests/test_sampling_return_signature.py [hunk 4]
```diff
@@ -0,0 +1,954 @@
...
+*
+* This Source Code may also be made available under the following Secondary
+* Licenses when the conditions for such availability set forth in the Eclipse
+* Public License, v. 2.0 are satisfied: GNU General Public License, version 2
+* with the GNU Classpath Exception which is
+* available at https://www.gnu.org/software/classpath/license.html.
+*
+* SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+********************************************************************************
+"""
```
Comment (PietropaoloFrisoni) [inline tests/jax_tests/test_sampling_return_signature.py]:
```suggestion
# ********************************************************************************
# * Copyright (c) 2026 the Qrisp authors
# *
# * This program and the accompanying materials are made available under the
# * terms of the Eclipse Public License 2.0 which is available at
# * http://www.eclipse.org/legal/epl-2.0.
# *
# * This Source Code may also be made available under the following Secondary
# * Licenses when the conditions for such availability set forth in the Eclipse
# * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
# * with the GNU Classpath Exception which is
# * available at https://www.gnu.org/software/classpath/license.html.
# *
# * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
# ********************************************************************************
```

File: tests/jax_tests/test_sampling_return_signature.py [hunk 5]
```diff
@@ -0,0 +1,954 @@
...
+
+r"""Comprehensive tests for the new pytree-aware return signature of
+:func:`~qrisp.jasp.sample`.
+
+Coverage matrix
+---------------
+
+============================ ======= ========= ============ ======
+Case                         JIT     Terminal  Non-JIT     Backend
+============================ ======= ========= ============ ======
```
Comment (PietropaoloFrisoni) [inline tests/jax_tests/test_sampling_return_signature.py:25]: This test summary is actually not bad. Maybe we can tell AI to create a similar test table for every test in the future (ideally once the test coverage is a little more consistent)

File: src/qrisp/jasp/program_control/ev.py [hunk 6]
```diff
@@ -208,6 +211,88 @@ def expectation_value_eval_function(*args, shots=0):
...
+            # Second pass: acc is a 1-tuple of shaped accumulators
+            return tuple(a + v for a, v in zip(acc, [decoded_values]))
+
+        def _make_init_acc(leaf_dtypes, leaf_shapes):
+            """Build a tuple of typed zero-arrays, one per leaf.
+
+            Each accumulator has the same shape as its leaf (running sum,
+            not per-shot storage).
+            """
+            return tuple(
```
Comment (PietropaoloFrisoni) [inline src/qrisp/jasp/program_control/ev.py]: I run our common friend Claude and found something interesting here (I think we can solve this by promoting `bool` to a summable `dtype`  before allocating the accumulator):

## `expectation_value()` silently returns a wrong result for `bool`-typed leaves

Found while reviewing the new typed-accumulator strategy in `ev.py`. This is a silent wrong answer.

### Root cause

`_make_init_acc` builds a per-leaf accumulator using the leaf's **native dtype** ([ev.py#L291-L294](https://github.com/eclipse-qrisp/Qrisp/blob/01927e37014c4825ced68bf8c7b35f46a5435df9/src/qrisp/jasp/program_control/ev.py#L291-L294)), and `_update_acc` accumulates via `acc + v` each shot ([ev.py#L261](https://github.com/eclipse-qrisp/Qrisp/blob/01927e37014c4825ced68bf8c7b35f46a5435df9/src/qrisp/jasp/program_control/ev.py#L261)). For a `bool` leaf, `bool + bool` in JAX/NumPy is **logical OR, not arithmetic addition**: it saturates at `True` on the first hit instead of counting. `QuantumBool.jdecoder` casts measurements to a genuine `dtype="bool"` array, so any `QuantumBool` inside a multi-value `expectation_value()` return hits this.

(`sample()` is *not* affected — it uses indexed `.at[i].set(v)`, not `+`, so it stores each shot rather than summing them.)

### Minimal repro (Qrisp)

```python
from qrisp import QuantumFloat, QuantumBool, h
from qrisp.jasp import expectation_value, jaspify

def kernel():
    qf = QuantumFloat(4)
    qb = QuantumBool()
    h(qf[0])   # E[qf] should be ≈0.5
    h(qb)      # E[qb] should be ≈0.5
    return qf, qb

@jaspify
def main():
    return expectation_value(kernel, shots=2000)()

print(main())
# Expected: (≈0.5, ≈0.5)
# Actual:   (≈0.5, ≈0.0005)   <- bool component wrong by ~1000x

File: src/qrisp/jasp/program_control/sampling.py [hunk 7]
```diff
@@ -121,8 +122,11 @@ def sample(sampling_kernel=None, shots=0, post_processor=None):
     Returns
     -------
     callable
-        A classical, Jax traceable function returning a jax array containing
-        the measurement results of each shot.
+        A classical, Jax traceable function.  For a kernel returning a single
+        value the result is a 1D ``jax.Array`` of length ``shots``.  For a
```
Comment (PietropaoloFrisoni) [inline src/qrisp/jasp/program_control/sampling.py]:
```suggestion
        value the result is a ``jax.Array`` of length ``shots``.  For a
```

Since in `test_single_non_scalar_array_jit` the shape is `(SHOTS, 2, 2)`

**Review by PietropaoloFrisoni — COMMENTED** [review]
Thanks @positr0nium , this is quite nice!

I also asked Claude to report the new lines not covered by tests (we should have a CI check for this). If we could test them as well, it would be great:

```
## Coverage gaps: newly introduced lines with no test exercising them

- `sampling.py:332-339` and `ev.py:249-255` — `except AttributeError` fallback when capturing leaf dtype/shape (triggers on a raw Python scalar inside a returned container; every test uses real JAX arrays, so it's dead code).
- `sampling.py:361` and `ev.py:278` — same fallback for the single non-scalar-leaf path; the surrounding branch runs under `test_single_non_scalar_array_jit`, but always with a real `.dtype`, so the fallback value itself is never produced.
- `ev.py`'s bool-leaf accumulation path (`ev.py:261`/`:283`, fed by `_make_init_acc` at `:291-294`) — executes on every run, but `test_bool_in_flat_tuple_jit` checks only `.dtype`, never the computed value, so its numeric correctness is unverified (see separate bug comment).
- `backend_sampling_interpreter.py:466-467` — `len(flat) == 1` re-collapse branch; no test returns a literal 1-tuple/1-list through `backend_sampler`.
- `backend_sampling_interpreter.py:464-469` — `dict` falls through unhandled (`isinstance` check only covers `tuple`/`list`); no `dict`-returning kernel is ever run through `backend_sampler`.
- `backend_sampling.py:447-450` — top-level pytree restoration in `_make_backend_sampler_wrapper`; every `@backend_sampler`-decorated test function returns a bare scalar or flat tuple at the top level, never `dict`/`list`/nested.
```

**Comment by purva-thakre** [comment]
> I also asked Claude to report the new lines not covered by tests (we should have a CI check for this).

We have this under the name `Coverage/report`?  You just have to scroll through the report to find the files changed in this PR. https://github.com/eclipse-qrisp/Qrisp/actions/runs/34227918171/job/102095811847?pr=749#step:6:16

**Comment by PietropaoloFrisoni** [comment]
Thanks @purva-thakre . It would be great if it could flag new source code lines with missing coverage as failures (same as `reviewdog` does now with `ruff` violations) and possibly point out the faulty line(s) in the files tab directly (we used codecov before and in most cases it was quite helpful), but we can discuss this in the future

**Comment by purva-thakre** [comment]
For `codecov`, we require a higher level of access to be able to connect it to the project. I have looked into this already. @bastibock will have to check if he has access; otherwise we reach out to the Eclipse foundation via their IT help desk. 

Feel free to implement the workflow you want to add for `reviewdog` and `coverage`. If I recall correctly, coverage was added before `reviewdog`. I think I am a bit burnt out by mostly working on CI changes. I want to work on the quantum stuff for a while at least.

**Comment by PietropaoloFrisoni** [comment]
> I think I am a bit burnt out by mostly working on CI changes. I want to work on the quantum stuff for a while at least

I completely understand. As far as I am concerned, the work that you did on the CI was extremely helpful for every developer working on qrisp (and, indirectly, beneficial for users as well)!

File: src/qrisp/jasp/program_control/ev.py [hunk 8]
```diff
@@ -208,6 +211,88 @@ def expectation_value_eval_function(*args, shots=0):
...
+                jnp.zeros(shape, dtype=dt) if dt is not None else jnp.zeros(shape)
+                for dt, shape in zip(leaf_dtypes, leaf_shapes)
+            )
```
[the remark this answers is not in the document]
Comment (positr0nium) [reply src/qrisp/jasp/program_control/ev.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/9180e0c151fc8e0be9c090f0e7b447ed14d80fb4

File: src/qrisp/jasp/interpreter_tools/interpreters/terminal_sampling_interpreter.py [hunk 9]
```diff
@@ -349,6 +349,19 @@ def sampling_body_eqn_evaluator(eqn, context_dic):
...
+        # match the expected number of outputs.
+        # - "array" mode produces a 2D (shots, n) array → n × 1D arrays
+        # - "ev"   mode produces a 1D (n,) array      → n × 0D scalars
+        if len(eqn.outvars) > 1 and len(decoded_meas_res) == 1:
+            flat = decoded_meas_res[0]
+            if hasattr(flat, "ndim"):
+                if flat.ndim == 2 and flat.shape[1] == len(eqn.outvars):  # noqa: PLR2004
+                    decoded_meas_res = [flat[:, i] for i in range(flat.shape[1])]
+                elif flat.ndim == 1 and flat.shape[0] == len(eqn.outvars):
+                    decoded_meas_res = [flat[i] for i in range(flat.shape[0])]
```
[the remark this answers is not in the document]
Comment (positr0nium) [reply src/qrisp/jasp/interpreter_tools/interpreters/terminal_sampling_interpreter.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/6cde8e5cf43897e7c7ce3f3a5fee7e35243f5aec

File: src/qrisp/jasp/program_control/sampling.py [hunk 10]
```diff
@@ -278,6 +273,109 @@ def sampling_eval_function(*args, tracerized_shots=0):
...
+                    return acc.at[i].set(decoded_values)  # true scalar
+
+                # Non-scalar leaf array — capture shape and retry
+                if not return_amount:
+                    leaf_dtype = getattr(decoded_values, "dtype", None)
+                    return_amount.append((struct, [leaf_dtype], [leaf_shape]))
+                raise _MultiReturnDetected()
+
+            # Second pass: acc is a 1-tuple of shaped accumulators
+            return tuple(a.at[i].set(v) for a, v in zip(acc, [decoded_values]))
```
[the remark this answers is not in the document]
Comment (positr0nium) [reply src/qrisp/jasp/program_control/sampling.py:362]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/d5a6f96f246296caa87db8ad00ca73494844860b

File: src/qrisp/jasp/program_control/sampling.py [hunk 11]
```diff
@@ -121,8 +122,11 @@ def sample(sampling_kernel=None, shots=0, post_processor=None):
     Returns
     -------
     callable
-        A classical, Jax traceable function returning a jax array containing
-        the measurement results of each shot.
+        A classical, Jax traceable function.  For a kernel returning a single
+        value the result is a 1D ``jax.Array`` of length ``shots``.  For a
+        kernel returning a container (``tuple``, ``list``, ``dict``, or nested
+        combinations thereof) each leaf is replaced by a 1D array preserving
+        its native dtype — e.g. ``{'a': bool_array, 'b': float_array}``.
```
[the remark this answers is not in the document]
Comment (positr0nium) [reply src/qrisp/jasp/program_control/sampling.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/fb2c7848e772aa13fce0864868aebd905b8dd100

File: tests/jax_tests/test_ev_return_signature.py (hunk 3 above)
Comment (positr0nium) [reply tests/jax_tests/test_ev_return_signature.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/fb2c7848e772aa13fce0864868aebd905b8dd100

File: tests/jax_tests/test_sampling_return_signature.py (hunk 4 above)
Comment (positr0nium) [reply tests/jax_tests/test_sampling_return_signature.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/fb2c7848e772aa13fce0864868aebd905b8dd100

File: src/qrisp/jasp/program_control/ev.py (hunk 6 above)
Comment (positr0nium) [reply src/qrisp/jasp/program_control/ev.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/9180e0c151fc8e0be9c090f0e7b447ed14d80fb4

File: src/qrisp/jasp/program_control/sampling.py (hunk 7 above)
Comment (positr0nium) [reply src/qrisp/jasp/program_control/sampling.py]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/fb2c7848e772aa13fce0864868aebd905b8dd100

File: documentation/source/general/changelog/changelog-dev.rst (hunk 1 above)
Comment (positr0nium) [reply documentation/source/general/changelog/changelog-dev.rst]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/9180e0c151fc8e0be9c090f0e7b447ed14d80fb4

File: documentation/source/general/changelog/changelog-dev.rst (hunk 2 above)
Comment (positr0nium) [reply documentation/source/general/changelog/changelog-dev.rst:32]: Fixed in https://github.com/eclipse-qrisp/Qrisp/pull/749/commits/9180e0c151fc8e0be9c090f0e7b447ed14d80fb4

File: src/qrisp/jasp/evaluation_tools/backend_sampling.py [hunk 12]
```diff
@@ -437,7 +437,17 @@ def eqn_evaluator(eqn, context_dic):
...
+        # only one of them), so the structure of the decorated function's return
+        # value is restored from out_shape, which is a pytree of
+        # ShapeDtypeStructs mirroring that structure. Without this, a function
+        # returning a container of values - say one list of X detectors and one
+        # list of Z detectors - would give the flattened values, leaving no way
+        # to tell which value belonged to which container.
+        if not isinstance(res, (tuple, list)):
+            res = [res]
+
+        return tree_unflatten(tree_structure(out_shape), list(res))
```
[the remark this answers is not in the document]
Comment (positr0nium) [reply src/qrisp/jasp/evaluation_tools/backend_sampling.py:450]: Thanks — I think this one is a false positive, caused by two similarly-named functions in the codebase.

`backend_sampling.py` imports `make_jaxpr` from `qrisp.jasp`, and that name resolves to **JAX's** `make_jaxpr`, not to Qrisp's `make_jaxpr_mod`. `centerclass.py:24` does `from jax import make_jaxpr`, and that is the symbol re-exported through the `qrisp.jasp` namespace. `make_jaxpr_mod` is a separate function, used only internally at `centerclass.py:1497`, and never exported under the name `make_jaxpr`.

The line you cited — `centerclass.py:1337`, `return closed_jaxpr, traced._out_tree` — is indeed inside `make_jaxpr_mod`, so the reasoning is right, just about the wrong function.

The distinction matters exactly as you describe:

```python
>>> f = lambda x: (x + 1, x * 2)

# JAX's make_jaxpr — what is actually imported here
>>> _, out_shape = make_jaxpr(f, return_shape=True)(1.0)
>>> out_shape
(ShapeDtypeStruct(shape=(), dtype=float64), ShapeDtypeStruct(shape=(), dtype=float64))
>>> tree_structure(out_shape).num_leaves
2                    # correct — a pytree OF ShapeDtypeStructs

# Qrisp's make_jaxpr_mod — what this comment assumes
>>> _, mod_tree = make_jaxpr_mod(f, return_shape=True)(1.0)
>>> mod_tree
PyTreeDef((*, *))
>>> tree_structure(mod_tree).num_leaves
1                    # the failure mode you describe
```

So `tree_structure(out_shape)` is correct as written, and passing `out_shape` directly would *introduce* the bug: `tree_unflatten` would get a `PyTreeDef` where it expects a pytree, and any multi-leaf return would raise `Too many leaves for PyTreeDef; expected 1`.

`test_multiple_sample_calls` already exercises a multi-leaf top-level return through this path and passes. I've also added `test_decorated_function_returns_dict`, `_nested_list` and `_single_element_container`, so dict, nested and single-leaf top-level returns are now covered too.

**Review by PietropaoloFrisoni — APPROVED** [review]
🚀

File: src/qrisp/jasp/program_control/ev.py
Comment (renezander90) [inline src/qrisp/jasp/program_control/ev.py:61]: The first argument is still called `state_prep` which is inconsistent with the documentation (`sampling_kernel`).

Comment (renezander90) [inline src/qrisp/jasp/program_control/ev.py:61]: The `return_dict` argument is not documented. What is its purpose? Why does it "keep legacy flat-array format"?

**Comment by renezander90** [comment]
```
from qrisp import *

def sampling_kernel(k):
    a = QuantumArray(QuantumFloat(2), shape=(2,))
    h(a)
    return a

@jaspify(terminal_sampling=True)
def main(k):

    samples = sample(sampling_kernel, shots = 10)(k)
    
    return samples

print(main(3))
```
> TypeError: Unsupported return type 'QuantumArray'. sample() supports tuple, list, and dict containers. Convert the return value to one of these types.

Why doesn't it support QuantumArrays?
